### PR TITLE
[FLINK-10289] [JobManager] Classify Exceptions to different category for apply different failover strategy

### PR DESCRIPTION
**Pull request by isunjin**
## What is the purpose of the change

We need to classify exceptions and treat them with different strategies. To do this, we propose to introduce the following Throwable Types, and the corresponding exceptions:

- NonRecoverable

   - We shouldn’t retry if an exception was classified as NonRecoverable
   - For example, NoResouceAvailiableException is a NonRecoverable Exception
   - Introduce a new Exception UserCodeException to wrap all exceptions that throw from user code
- PartitionDataMissingError
   - In certain scenarios producer data was transferred in blocking mode or data was saved in persistent store. If the partition was missing, we need to revoke/rerun the produce task to regenerate the data.
   - Introduce a new exception PartitionDataMissingException to wrap all those kinds of issues.
- EnvironmentError
   - It happened due to hardware, or software issues that were related to specific environments. The assumption is that a task will succeed if we run it in a different environment, and other task run in this bad environment will very likely fail. If multiple task failures in the same machine due to EnvironmentError, we need to consider adding the bad machine to blacklist, and avoiding schedule task on it.
   - Introduce a new exception EnvironmentException to wrap all those kind of issues.
- Recoverable
   - We assume other issues are recoverable.

## Brief change log

  -  *Add exception types*
  -  *Add a class to classify exceptions*
  -  *Unittests*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added test that validates SuppressRestartsException was a NonRecoverable Exception*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (*no*)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? YES, Document is [(Here)](https://docs.google.com/document/d/1FdZdcA63tPUEewcCimTFy9Iz2jlVlMRANZkO4RngIuk/edit?spm=a1zcr.8293797.0.0.3b116385Btb5sf)

### CODE REVIEW

**Comment by StephanEwen** [comment]
I really like this idea in principle.

Instead of attaching an enum classifier, could we solve this by exception inheritance? Having exceptions that represent the categories and then making the specific exceptions subclasses of the category exceptions?

**Comment by isunjin** [comment]
> I really like this idea in principle.
> 
> Instead of attaching an enum classifier, could we solve this by exception inheritance? Having exceptions that represent the categories and then making the specific exceptions subclasses of the category exceptions?

Thanks Stephan, use exception inheritance certainly a good way, however:
- enum explicitly tell that there are four types of exceptions, while exception is more scattered. 
- Java doesn't support multi-inherits that will make subclasses a little harder but we can also use Java interface or Java annotation 
- There is a high level class need to consume this classification logic, so the ThrowableClassifier here is simply the logic, eg: if we use annotation, we need a logic to extract the annotation.

so i would propose to mix those ideas together, keep the classify logic and use Java annotation/interface, @StephanEwen let me know what you think.

Jin

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+package org.apache.flink.runtime.throwable;
+
+/**
+ * */
+public enum ThrowableType {
+
+	/**
+	 * the issue indicate it will not success even retry, such as a DivideZeroException.
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java]: this indicates error that would not succeed even with retry, such as DivideZeroExeception. No failover should be attempted with such error. Instead, the job should fail immediately.

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	 * for this kind of exception, we shouldn't consider failover, or we should fail the job
+	 */
+	NonRecoverable,
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java]: NonRecoverableError

to be more consistent with other types

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+	/**
+	 * indicate data consumption error, that we should revoke the producer.
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java]: data consumption error, which indicates that ...

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+	 * */
+	PartitionDataMissingError,
+
+	/**
+	 * this indicate errors such us Hardware error, service issue, that we should consider blacklist the machine.
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java]: this indicates error related to running environment, such as ..., in which case we should consider...

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+	 * */
+	EnvironmentError,
+
+	/**
+	 * this indicate other errors that recoverable.
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableType.java]: this indicates other error that is recoverable

File: flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableClassifier.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import org.apache.flink.runtime.jobmanager.scheduler.NoResourceAvailableException;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * given a exception do the classification.
+ */
+public class ThrowableClassifier {
```
Comment (yingdachen) [inline flink-runtime/src/main/java/org/apache/flink/runtime/throwable/ThrowableClassifier.java]: i think we can explore interface/annotation approach

**Comment by isunjin** [comment]
@StephanEwen , i have changed the implementation with java annotation, could you take a look?

**Comment by StefanRRichter** [comment]
@isunjin I also like the general idea, but not so much the current implementation. One disadvantage I see with the annotation is that they are not propagated in inheritance hierarchies and you always need to remember adding them in such a case. I also wonder where and how this classifier will be invoked later. Maybe an alternative approach could be to have a fixed set of `TaskException`s for the cases, and before leaving the TM, all exceptions are wrapped as cause into their matching `TaskException`  to classify them for the JM's recovery approach. However, the "best" approach might also depend on the intended use, which is not clear for me from the PR and that is why I am asking.

**Comment by isunjin** [comment]
@StefanRRichter Thanks for review and you make a very good point, but annotation can be propagate with a '@Inherited' annotation, i have modified the code with unittest.

for the consumer of this class, i would like to split the commit to make it simple, it will be in another PR, that basically implement the logic mentioned in the design doc.

**Comment by StefanRRichter** [comment]
@isunjin I think this looks good now. I wonder if we should rename the type `Other` to something like `RecoverableFailure`?

**Comment by isunjin** [comment]
@StefanRRichter Thank you, pushed the change.

**Pull request closed without merging**